### PR TITLE
Add initial pyodide support

### PR DESCRIPTION
**Pull request by alcarney**
## Description

With the recent releases of [github.dev](https://github.dev/openlawlibrary/pygls) and [vscode.dev](https://vscode.dev/github/openlawlibrary/pygls) I wanted to see if it was possible to use [Pyodide](https://pyodide.org/en/stable/) to have a `pygls` powered language server running in a web browser. - Turns out it is! :smile: 

It's worth noting that due to the `multiprocessing` module [not being available](https://pyodide.org/en/stable/usage/wasm-constraints.html#included-but-not-working-modules) in Pyodide, `@server.thread()` methods will not work in the browser, but so far everything else appears to work.

This PR makes the few tweaks required to get `pygls` running in a browser context (see commit messages for details) as well as adding a new example extension that demonstrates how to get a simple language server up and running in a web version of VSCode.

This isn't quite ready to be merged yet as ~`mypy` isn't happy about the import from Pyodide's `js` module and~ I should probably update the docs with some notes on Pyodide, but I thought I'd open it now so that there's an opportunity for feedback :smile: 

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/pygls/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/
[contributors]: https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTORS.md

### CODE REVIEW

File: pygls/protocol.py
Comment (alcarney) [inline pygls/protocol.py:383]: This requires there be a JavaScript `post_message` method be defined on the server's web worker.
See [here](https://github.com/openlawlibrary/pygls/blob/f6d1f9e3947a9175b5998424db3da9ebcbdefc22/examples/fountain-extension/src/server.ts#L34-L41)

**Comment by willemkokke** [comment]
I would love to help out on this one, I want a to hook up a python LS running in pyodide to a monaco editor using [Monaco-Language](https://github.com/TypeFox/monaco-languageclient), and I've identified pygls as the path of least resistance.

Thanks for the pull request @alcarney, I was just about to rip multiprocessing out myself when I thought to check open PRs ;)

If there's anything I can do to help this land please let me know!

**Comment by dgreisen** [comment]
@willemkokke (and everybody else! :) - @danixeee doesn't have as much time for pygls right now as we'd all like. Open Law Library is looking for someone interested in a part-time (~40hrs/mo) paid position maintaining pygls. We are also looking for volunteer community maintainers. If you know anybody who might be interested, please let me know (dgreisen@openlawlib.org).

**Comment by alcarney** [comment]
I'd be interested in helping out as a volunteer where I can :)

**Comment by tombh** [comment]
This is such a cool PR! Thank you. We want to get this merged pretty much as is. Then we can start another issue or PR about adding some Pyodide-specific tests to make sure this keeps working. I've just added a couple of comments, if we can clear that up then LGTM.

**Comment by alcarney** [comment]
>  I've just added a couple of comments, if we can clear that up then LGTM.

Hmm.. the only comment I can see is [this one](https://github.com/openlawlibrary/pygls/pull/218#issuecomment-1141338221) - could you share a link to one of the others?

File: pygls/__init__.py
Comment (alcarney) [reply pygls/__init__.py:23]: I see them now.... not sure what happened there.

> Is there a chance of a false positive here? The only way that pyodide can be part of sys.modules is if the parent application explicitly imports it right?

I'm assuming Pyodide is doing a little bit of magic to make it available? 
It is the suggested method in the project's [FAQ](https://pyodide.org/en/stable/usage/faq.html#how-to-detect-that-code-is-run-with-pyodide) for detecting the runtime.

File: pygls/server.py
Comment (alcarney) [reply pygls/server.py:270]: I believe `"ThreadPool"` will be resolved to `ThreadPool` by tools like mypy during type checking.

If I remember rightly, the issue is that importing `ThreadPool` from `multiprocessing` crashed under Pyodide so [this line](https://github.com/openlawlibrary/pygls/pull/218/files/629e6a07982b47a308c35842b4a430461bc88332#diff-073264baeffc39cb93e2bc9bd897133fb5d3ba00db0dd0bb073af883c15b6b61R36) disables the import (multiprocessing is [ not available](https://pyodide.org/en/stable/usage/wasm-constraints.html#included-but-not-working-modules) under Pyodide anyway) meaning we have to switch to using quotes

File: pygls/__init__.py
Comment (tombh) [reply pygls/__init__.py:23]: Ok, the suggested method must be pretty tried and tested then. Thanks

File: pygls/server.py
Comment (tombh) [reply pygls/server.py:270]: That makes sense. So would this work too?
```python
if not IS_PYODIDE:
    from multiprocessing.pool import ThreadPool
else:
   ThreadPool = "ThreadPool"

.
.

def thread_pool(self) -> ThreadPool:
```

Comment (alcarney) [reply pygls/server.py:270]: Not sure…. I don’t know the specifics of typing enough to say either way.
Is there a particular reason for doing it that way? Can’t say it’s something I’ve seen before

Comment (alcarney) [reply pygls/server.py:270]: So I tried the above form out and `mypy` didn't like it.
```
pygls/server.py:39: error: Cannot assign to a type
pygls/server.py:39: error: Incompatible types in assignment (expression has type "str", variable has type "Type[ThreadPool]")
```

**Comment by tombh** [comment]
Thanks. So what about this?:
```python
if IS_PYODIDE:
    ThreadPoolType = "ThreadPool"
else:
    from multiprocessing.pool import ThreadPool
    ThreadPollType = ThreadPool
.
.

def thread_pool(self) -> ThreadPoolType:
```

I don't really know enough about mypy either. It just seems a shame to lose typing information when not using Pyodide

**Comment by davelopez** [comment]
Maybe you can use `typing.TYPE_CHECKING` [see docs here](https://mypy.readthedocs.io/en/stable/runtime_troubles.html#annotation-issues-at-runtime).

**Comment by alcarney** [comment]
> Thanks. So what about this?:

That also fails - but with slightly different error messages
```
pygls/server.py:41: error: Incompatible types in assignment (expression has type "Type[ThreadPool]", variable has type "str")
pygls/server.py:274: error: Variable "pygls.server.ThreadPoolType" is not valid as a type
```

>  It just seems a shame to lose typing information when not using Pyodide

I wouldn't say that the typing information is lost... VSCode understands the type fine even when it is quoted
image

And it's even possible to get the proper type at runtime
```python
>>> import inspect
>>> import typing
>>> from pygls.server import Server
>>> print(inspect.getsource(Server.thread_pool.fget))
    @property
    def thread_pool(self) -> "ThreadPool":
        """Returns thread pool instance (lazy initialization)."""
        if not self._thread_pool:
            self._thread_pool = ThreadPool(processes=self._max_workers)

        return self._thread_pool

>>> typing.get_type_hints(Server.thread_pool.fget)
{'return': <class 'multiprocessing.pool.ThreadPool'>}
>>> 
```
> Maybe you can use typing.TYPE_CHECKING

Thanks for the suggestion! If we used `typing.TYPE_CHECKING` it could be simplified to 

```python
if typing.TYPE_CHECKING:
    from multiprocessing.pool import ThreadPool
.
.

def thread_pool(self) -> "ThreadPool":
```

**Comment by tombh** [comment]
Thanks for all that info! Ok, so quoting is just another valid way of expressing a type, it doesn't mean the type is a literal.

So my worry is that somebody in the future will see this break in convention (of expressing the type with quotes when all other types are without quotes) and innocently remove the quotes, breaking Pyodide support (which doesn't have tests yet). More broadly speaking, I think it's best to strive to maintain the precedent that the core code dictates to non-core features. In that sense, here Pyodide should bend to the core, not the other way around. So quoting `"ThreaPool"` certainly isn't a big deal, but its break in precedent could be interpreted as a big deal.

I like the `IS_PYODIDE` guards, it'd be perfect if that last little issue of the type could live behind a guard too.

So here are the options I see:
  * At the very least that line should be commented, explaining the surprising break in precedent.
  * Better would be if the type had something in its name eg; `"ThreadPoolAliasType"` (comments are beyond the reach of tests and linters, and tend to get stale) or something that provided a clue for the quoting.
  * A complete test suite for Pyodide running on CI.
  * What about guarding the whole `thread_pool()` function(s)!? Eg;
```python
if IS_PYODIDE:
    @property
    def thread_pool(self) -> ThreadPool:
        ...

    @property
    def thread_pool_executor(self) -> ThreadPoolExecutor:
        ...
```

**Comment by alcarney** [comment]
> What about guarding the whole `thread_pool()` functions?

I like that idea, I imagine it should work and makes it a lot clearer that they can't be relied on in a Pyodide context as they simply won't exist.

> A complete test suite for Pyodide running on CI.

Do you want to tackle that as part of this PR? After some very hasty googling I see a couple of options
- We can try using [@vscode/test-web](https://code.visualstudio.com/api/extension-guides/web-extensions#test-your-web-extension-in-a-browser-using-vscodetestweb) to test the example VSCode extension
- Pyodide looks like it uses selenium and/or playwright to test itself, I don't see an obvious reusable component in the repo but we might be able to use their test suite as inspiration

**Comment by tombh** [comment]
Personally I think the tests can be worked on afterwards, once the PR is merged. We can make a separate issue for that as not having tests for a feature should be considered a bug. You've already done so much cool work here, it's high time it made it into master so others can kick its tyres.

So do you want to put that guard at the function-level, then LGTM?

**Review by tombh — APPROVED** [review]

**Comment by tombh** [comment]
Merged! 🥳 I'll add an issue to add tests...

**Pull request merged**
